src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs

```diff
@@ -52,34 +55,38 @@ public MapReduceResultsStore(ulong reduceKeyHash, MapResultsStorageType type, Tr
             }
         }
 
+        private readonly HashSet<string> _alreadyInitializedTrees = new HashSet<string>();
+
         private void InitializeTree(bool create)
         {
             var treeName = ReduceTreePrefix + _reduceKeyHash;
             var options = _tx.LowLevelTransaction.Environment.Options.RunningOn32Bits ? TreeFlags.None : TreeFlags.LeafsCompressed;
             Tree = create ? _tx.CreateTree(treeName, flags: options) : _tx.ReadTree(treeName);
 
-            ModifiedPages = new HashSet<long>();
-            FreedPages = new HashSet<long>();
+            if (!_alreadyInitializedTrees.Contains(treeName))
+            {
+                Tree.PageModified += (page, flags) =>
+                {
+                    if ((flags & PageFlags.Overflow) == PageFlags.Overflow)
+                        return;
 
-            _mapReduceContext.PageModifiedInReduceTree += page => FreedPages.Remove(page);
```

Comment:
Why did you remove `PageModifiedInReduceTree`? 
It made that  `RavenDB_11643.PageModificationInAnyTreeMustRemoveItFromListOfFreedPagesInAllStores` failed (http://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.1/2220/)

Please check https://github.com/ravendb/ravendb/pull/7690/ it was added there to handle edge case

Reply:
Back to the drawing board then. I need to understand what is the rationale of that edge case. Can you elaborate?

Reply:
So yes, my fix was: if a page is modified (in any tree), let's remove it from _all_ `FreedPages` collections (for every `MapReduceResultsStore`)

Reply:
Is not reusing pages at all a viable strategy?

Reply:
Shared `FreedPages` makes sense to me. Since this code:

https://github.com/ravendb/ravendb/blob/436d63cd0b9b42473d7dc0945b8f4ebd2be6b2a3/src/Raven.Server/Documents/Indexes/MapReduce/ReduceMapResultsBase.cs#L332-L336

Doesn't need to be run in "contex" of any particular tree, we can run it before starting processing any tree (`HandleTreeReduction`) just once. That should be fine I think (tests will validate that for sure).

Reply:
One potential problem, `HandleTreeReduction` might be called recursively

https://github.com/ravendb/ravendb/blob/436d63cd0b9b42473d7dc0945b8f4ebd2be6b2a3/src/Raven.Server/Documents/Indexes/MapReduce/ReduceMapResultsBase.cs#L406-L427

So maybe we'll be better of by calling `table.DeleteByKey(pageNumberSlice); ` for shared `FreedPages` _after_ processing all trees